protex/system.py

```diff
@@ -0,0 +1,215 @@
+
+        assert len(set(list(self.charge_sets.keys()))) == 2
+        # calculate initial charge
+        intial_charge = int(
+            np.round(
+                sum(
+                    [
+                        self.nonbonded_force.getParticleParameters(idx)[0]._value
+                        for idx in self.atom_idxs
+                    ]
+                ),
+                4,
+            )
+        )
+        assert intial_charge in list(self.charge_sets.keys())
+        self.record_charge_state.append(intial_charge)
+
+    def get_current_charge(self) -> int:
+        return self.record_charge_state[-1]
+
+    def get_current_charges(self) -> list:
+        return self.charge_sets[self.record_charge_state[-1]]
+
+    def get_inactive_charges(self) -> list:
+        """
+        get_inactive_charges retrieves the charges that are not active in the residue
+
+        Returns
+        -------
+        list
+            list of charges
+
+        Raises
+        ------
+        RuntimeError
+            if none found
+        """
+        current_charge = self.get_current_charge()
+        for charge in self.charge_sets:
+            if charge != current_charge:
+                return self.charge_sets[charge]
+
+        else:
+            raise RuntimeError()
+
+    def set_new_charge(self, new_charges: list) -> None:
+        from simtk import unit
+
+        for new_charge, idx in zip(new_charges, self.atom_idxs):
+            (
+                _,
+                old_sigma,
+                old_epsilon,
+            ) = self.nonbonded_force.getParticleParameters(idx)
+            self.nonbonded_force.setParticleParameters(
+                idx, new_charge * unit.elementary_charge, old_sigma, old_epsilon
+            )
+
+
+class IonicLiquidSystem:
+    """
+    This class defines the full system, performs the MD steps and offers an
+    interface for protonation state updates.
+    """
+
+    def __init__(self, simulation, templates: IonicLiqudTemplates) -> None:
+        self.system = simulation.system
+        self.topology = simulation.topology
+        self.simulation = simulation
+        self.templates = templates
+        for force in self.system.getForces():
+            if (type(force).__name__) == "NonbondedForce":
+                self.nonbonded_force = force
+
+        self.residues = self._set_initial_states()
+
+    def _set_initial_states(self) -> list:
+        """
+        set_initial_states For each ionic liquid residue in the system the protonation state
+        is interfered from the provided openMM system object and the protonation site is defined.
+        """
+        residues = []
+
+        for r in self.topology.residues():
+            name = r.name
+            if name in self.templates.names:
+                name_of_paired_ion = (
+                    self.templates.get_residue_name_for_other_charged_state(name)
+                )
+                residues.append(
+                    Residue(
```

Comment:
Just for my understanding: for a residue of the Residue class all methods defined by OpenMM as well as our self-written ones of the Residue Class can be called?

Reply:
the Residue class includes an openMM residue instance, and we define a few methods that are helpful for our objective.